### PR TITLE
IGNITE-13659 Cache encryption key rotation (documentation)

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+=== Using XML configuration to limit re-encryption rate
+[source, xml]
+----
+include::code-snippets/xml/tde.xml[tags=ignite-config;!discovery;!encryption;!cache;!discovery, indent=0]
+----
+
+=== Using CLI to control re-encryption
+
+The `control.sh|bat` script provides the ability to change the re-encryption rate, suspend and resume background re-encryption at runtime.
```
Comment (Silberfuchs1) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:120]: '...as well as suspend...'

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+NOTE: After the node restarts, the suspended background re-encryption will continue automatically, and the rate limit will be set from the local XML configuration (if any) or set to unlimited rate (default).
```
Comment (Silberfuchs1) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:122]: '...re-encryption will continue automatically' --> '...re-encryption will be continued automatically'
'...and the rate limit will be set to 'unlimited' (by default), or taken from the local XML configuration (if any).'

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+[source,shell]
+----
+# View the re-encryption status of the cache group.
```
Comment (Silberfuchs1) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:126]: 'View the cache group re-encryption status.'

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+# Resume (suspended) re-encryption of the cache group.
+control.sh|bat --encryption resume_reencryption cacheGroupName
+
+# View the re-encryption rate limit.
+control.sh|bat --encryption reencryption_rate_limit
+
+# Set the re-encryption rate limit to 2.5 MB/s.
+control.sh|bat --encryption reencryption_rate_limit 2.5
+
+# Set re-encryption to an unlimited rate.
```
Comment (Silberfuchs1) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:141]: 'Set re-encryption rate to 'unlimited' ('0').'

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+control.sh|bat --encryption reencryption_rate_limit 0
+----
+
+The re-encryption status can also be obtained using JMX metrics, which are described in the link:monitoring-metrics/new-metrics#cache-groups[Cache group metrics] section.
```
Comment (Silberfuchs1) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:145]: '...using JMX metrics described in the link:monitoring-metrics/new-metrics#cache-groups[Cache group metrics] section.'

**Review by Silberfuchs1 — CHANGES_REQUESTED** [review]
There are some minor issues. See my inline comments.

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+= Cache encryption key rotation
+
+== Overview
+
+The cache group encryption key is used to encrypt cache data on disk.
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Fixed, thanks.

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+When a new cache starts, a new encryption key is generated and propagated to all server nodes in the cluster. So each node for the same cache group has the same cache encryption key.
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Changed to "When a user creates a new encrypted cache"

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+To learn more see the link:security/tde[Transparent Data Encryption] page.
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Fixed, thanks.

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+Ignite 2.10 introduces a feature for changing the cache encryption key.
+It allows users to change the encryption key of the cache group and re-encrypt existing data at runtime.
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Fixed, thanks.

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+Rotation of the cache encryption key is required if this key has been compromised or at the end of the crypto period (key validity period).
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Fixed, thanks.

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+|reencryptionBatchSize | 100 | The number of pages that is scanned during re-encryption under checkpoint lock.
+|===
+
+=== Using XML configuration to limit re-encryption rate
+[source, xml]
+----
+include::code-snippets/xml/tde.xml[tags=ignite-config;!discovery;!encryption;!cache;!discovery, indent=0]
+----
+
+=== Using CLI to control re-encryption
```
[the remark this answers is not in the document]
Comment (xtern) [reply docs/_docs/security/cache-encryption-key-rotation.adoc]: Changed to "Using CLI to control re-encryption process"

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/TDE.java [hunk 12]
```diff
@@ -58,6 +58,13 @@ void configuration() {
         IgniteFuture<Void> future = ignite.encryption().changeMasterKey("newMasterKeyName");
         //end::master-key-rotation[]
 
+        //tag::cache-group-key-rotation[]
+        // Starts cache group encryption key change process.
+        // This future will be completed when the new encryption key is set for writing on
```
Comment (Nikita-tech-writer) [inline docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/TDE.java:63]:
```suggestion
        // This feature will be completed when the new encryption key is set for writing on
```

File: docs/_docs/monitoring-metrics/new-metrics.adoc [hunk 13]
```diff
@@ -129,6 +129,8 @@ Register name: `cacheGroups.{group_name}`
 |StorageSize |long|    Storage space allocated for group, in bytes.
 |TotalAllocatedPages |long|    Cache group total allocated pages.
 |TotalAllocatedSize  |long|    Total size of memory allocated for group, in bytes.
+|ReencryptionBytesLeft |long| The number of bytes left for re-ecryption.
```
Comment (Nikita-tech-writer) [inline docs/_docs/monitoring-metrics/new-metrics.adoc:132]:
```suggestion
|ReencryptionBytesLeft |long| The number of bytes left for re-encryption.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+= Cache encryption key rotation
+
+== Overview
+
+Cache group encryption key is used to encrypt cache data on disk.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:19]:
```suggestion
Cache group encryption key is used to encrypt cache data on the disk.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+When a user creates a new encrypted cache, a new encryption key is generated and propagated to all server nodes in the cluster. So, each node for the same cache group has the same cache encryption key.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:20]:
```suggestion
When a user creates a new encrypted cache, a new encryption key is generated and propagated to all server nodes in the cluster. So, each node has the same cache encryption key for the same cache group.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+See link:security/tde[Transparent Data Encryption] to learn more.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:21]:
```suggestion
See link:security/tde[Transparent Data Encryption] for more detail.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+Ignite 2.10 introduces a feature for changing the cache encryption key.
+It allows users to change the cache group encryption key and re-encrypt existing data at runtime.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:24]:
```suggestion
It allows to change the cache group encryption key and re-encrypt existing data at runtime.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+Rotation of the cache encryption key is required if this key has been compromised or if the crypto period (key validity period) has ended.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:26]:
```suggestion
Rotation of the cache encryption key is required when the key is compromised or the crypto period (key validity period) is ended.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+The process of changing the cache encryption key includes two sequential stages:
+
+1. Rotate cache group key. This process adds a new encryption key for the specified cache group(s) on each server node and sets it to write new data.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:30]:
```suggestion
1. Rotate cache group key. This process adds a new encryption key for the specified cache group or groups on each server node and sets it to write new data.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+    Node join during this stage is prohibited and will be rejected.
+
+2. Re-encrypt existing (archived) cache data with the new encryption key.
+
+The second stage can take a long time. It depends on the amount of existing data. During this period the old key will be kept to read the archived data.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:36]:
```suggestion
The second stage can take a while. It depends on the amount of existing data. During this period, the old key is kept to read the archived data.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+To understand what key the data is encrypted with, each encryption key has an _identifier_. By default it is equal to zero. The identifier of the new key increases with each rotation.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:37]:
```suggestion
To understand what key the data is encrypted with, each encryption key has an _identifier_. By default, it is equal to zero. The identifier value of the new key increases with each new rotation.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+The encryption key (as well as encryption key ID) is the same for all nodes in a cache group.
+
+NOTE: Secondary rotation of the cache encryption key is possible only after a complete change of the encryption key for a cache group (both stages).
+
+== Prerequisites
+
+The cluster should be active.
+
+== Changing the encryption key
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:46]:
```suggestion
== Changing the Encryption Key
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+Ignite provides the ability to change the cache encryption key using the following interfaces:
+
+- link:#command-line-tool[command line tool]
+- link:#jmx[JMX]
+- link:#from-code[from code]
+
+=== Command line tool
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:54]:
```suggestion
=== Command Line Tool
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+
+You can also change the cache encryption key via the `EncryptionMXBean` interface:
+
+[cols="1,1",opts="header"]
+|===
+|Method | Description
+|changeCacheGroupKey(String cacheOrGrpName) | Starts cache encryption key change process.
+|===
+
+=== From code
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:78]:
```suggestion
=== From Code
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+
+The cache encryption key change process can also be managed programmatically:
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:80]:
```suggestion
The cache encryption key change process can also be managed directly in the code:
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+--
+tab:Java[]
+
+[source, java]
+----
+include::{javaCodeDir}/TDE.java[tags=cache-group-key-rotation, indent=0]
+----
+--
+
+== Managing re-encryption
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:92]:
```suggestion
== Managing Re-encryption
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+
+Re-encrypting existing data can take a long time. This is a fault-tolerant operation that automatically continues after a node restart.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:94]:
```suggestion
Re-encrypting existing data can take a while. This is a fault-tolerant operation that automatically continues after a node restart.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+The previous encryption key is automatically removed when all local partitions are encrypted with the new key, and the last link:persistence/native-persistence#write-ahead-log[Write-Ahead Log] segment, which may contain entries encrypted with the previous key, is removed from disk.
+
+NOTE: Re-encryption uses link:persistence/native-persistence#write-ahead-log[Write-Ahead Log] for physical recovery and may affect performance of cache operations.
+
+There are several options to manage the performance impact of re-encryption:
+
+* Limit the re-encryption rate using a configuration parameter or CLI at runtime.
+* Temporarily suspend re-encryption using CLI command.
+
+Ignite 2.10 introduces a new configuration section `EncryptionConfiguration`, that is part of `DatastorageConfiguration`.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:104]:
```suggestion
Ignite 2.10 introduces a new configuration section `EncryptionConfiguration`, that is a part of `DatastorageConfiguration`.
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+[cols="1,1,1",opts="header"]
+|===
+|Property | Default value | Description
+|reencryptionRateLimit | 0 (unlimited) | Re-encryption rate limit in megabytes per second.
+|reencryptionBatchSize | 100 | The number of pages scanned during re-encryption under checkpoint lock.
+|===
+
+=== Using XML configuration to limit the re-encryption rate
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:112]:
```suggestion
=== Using XML Configuration to Limit the Re-encryption Rate
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+[source, xml]
+----
+include::code-snippets/xml/tde.xml[tags=ignite-config;!discovery;!encryption;!cache;!discovery, indent=0]
+----
+
+=== Using CLI to control re-encryption process
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:118]:
```suggestion
=== Using CLI to Control Re-encryption Process
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+
+The `control.sh|bat` script provides the ability to change the re-encryption rate as well as suspend and resume background re-encryption at runtime.
+
+NOTE: After the node restarts, the suspended background re-encryption will be continued automatically, and the rate limit will be set to 'unlimited' (by default), or taken from the local XML configuration (if any).
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:122]:
```suggestion
NOTE: After the node restarts, the suspended background re-encryption is continued automatically, and the rate limit is set to 'unlimited' (by default), or taken from the local XML configuration (if any).
```

File: docs/_docs/security/cache-encryption-key-rotation.adoc [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+control.sh|bat --encryption reencryption_rate_limit
+
+# Set the re-encryption rate limit to 2.5 MB/s.
+control.sh|bat --encryption reencryption_rate_limit 2.5
+
+# Set re-encryption rate to 'unlimited' ('0').
+control.sh|bat --encryption reencryption_rate_limit 0
+----
+
+The re-encryption status can also be obtained using JMX metrics described in the link:monitoring-metrics/new-metrics#cache-groups[Cache group metrics] section.
```
Comment (Nikita-tech-writer) [inline docs/_docs/security/cache-encryption-key-rotation.adoc:145]:
```suggestion
The re-encryption status can be also obtained using JMX metrics described in the link:monitoring-metrics/new-metrics#cache-groups[Cache group metrics] section.
```

**Review by Nikita-tech-writer — CHANGES_REQUESTED** [review]
Please see my suggested changes to the PR.

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/TDE.java (hunk 12 above)
Comment (xtern) [reply docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/TDE.java:63]: Thanks, but this is not a typo, it refers specifically to "future" and not a "feature"

**Comment by xtern** [comment]
@Nikita-tech-writer, I fixed all the mentioned issues. Do you approve these changes?

**Review by Mmuzaf — APPROVED** [review]

**Pull request merged**